robot/impl/resource_manager.go

```diff
@@ -410,11 +412,15 @@ func (manager *resourceManager) mergeResourceRPCAPIsWithRemote(r robot.Robot, ty
 }
 
 func (manager *resourceManager) closeResource(ctx context.Context, res resource.Resource) error {
-	allErrs := res.Close(ctx)
+	// Avoid hangs in Close/RemoveResource with resourceCloseTimeout.
+	closeCtx, cancel := context.WithTimeout(ctx, resourceCloseTimeout)
+	defer cancel()
+
+	allErrs := res.Close(closeCtx)
 
 	resName := res.Name()
 	if manager.moduleManager != nil && manager.moduleManager.IsModularResource(resName) {
-		if err := manager.moduleManager.RemoveResource(ctx, resName); err != nil {
+		if err := manager.moduleManager.RemoveResource(closeCtx, resName); err != nil {
```

Comment:
Modular resources will not necessarily "respect" this `closeCtx` either, but we'll get a `context.DeadlineExceeded` error here (tested locally) when a modular resource hangs in its `Close`/`close`/destructor. I think that's fine.

Reply:
if we fail, do we still remove from the graph?

Reply:
We should, yes. By this point in time, we've `RemoveMarked` from the resource graph, which returns a slice of resources to call `closeResource` on.